Step2 문자열 덧셈 계산기 구현

Ganjinajae opened the pull request:
안녕하세요. 지영진입니다.
문자열 덧셈 계산기 구현하여 pr올립니다.

“//”와 “\n” 문자 사이에 커스텀 구분자를 지정하여 계산하는 요구사항에 대한 커밋을 못해서
숫자 이외의 값 또는 음수를 전달했을 때의 커밋과 같이 들어가있습니다.

참고부탁드립니다.
감사합니다.

youngjin-j committed:
feat: 빈 문자열 또는 null 값 입력할 경우 0 반환

youngjin-j committed:
feat: 숫자 하나를 문자열로 입력할 경우 해당 숫자 반환

youngjin-j committed:
feat: 쉼표 구분자 있을 때 두 숫자의 합 반환

youngjin-j committed:
feat: 쉼표 구분자 외에 쌍따옴표 구분자 사용하여 계산하기

* 메서드 분리 시행

youngjin-j committed:
feat: 숫자 이외의 값 또는 음수를 전달할 경우 RuntimeException 예외 throw

* if 문 분기 필요 없는 것 정리
* 필요 없는 변수 정리

seokhongkim commented on src/test/java/study/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,53 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class StringAddCalculatorTest {
```
테스트 잘 작성해주셨습니다. 👍

seokhongkim commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private static String[] splitInput(String input) {
+        String[] numbers = input.split(",|:");
```
의미를 나타낼 수 있도록 상수로 관리하고 이름을 정해보면 어떨까요?

Ganjinajae replied:
넵 확인하였습니다.

seokhongkim commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,52 @@
+public class StringAddCalculator {
+
+    public static int splitAndSum(String input) {
+        if (isNullOrEmptyString(input)) return 0;
```
기본 값임을 나타낼 수 있도록 `0` 도 상수로 관리해보시면 좋겠네요 :)

Ganjinajae replied:
넵 확인하였습니다.

seokhongkim commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,52 @@
+        if (isNullOrEmptyString(input)) return 0;
+        String[] numbers = splitInput(input);
+        validatePositiveNumbers(numbers);
+        if (isOnlyNumber(numbers)) return 1;
```
`numbers` 에 값이 하나만 있는 경우, 그 값을 반환해야 하지 않을까요?
테스트가 성공한 이유를 보니 테스트에서 1을 입력으로 전달해서 그런 것 같습니다.

Ganjinajae replied:
넵 확인하였습니다.

seokhongkim commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,52 @@
+public class StringAddCalculator {
+
+    public static int splitAndSum(String input) {
+        if (isNullOrEmptyString(input)) return 0;
```
if, for 구문에서 중괄호를 사용하시는 경우와 사용하지 않는 경우가 존재하는 것 같습니다.
하나로 통일하고 컨벤션을 맞춰보시면 좋겠습니다. :)

Ganjinajae replied:
넵. 통일해보겠습니다.

seokhongkim commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringAddCalculator {
```
클래스가 두가지 역할을 하고 있는 것 같아요.

1. 입력받은 문자열을 파싱한다. (숫자로 변환 포함)
2. 숫자를 더한 값을 계산한다.

각 역할을 클래스로 나눠보시면 어떨까요?

Ganjinajae replied:
아.. 미션에서 나와있는 테스트코드 대로 하다보니 별 생각없이 한 클래스로 한 것 같은데요.
한 번 나눠보겠습니다.

seokhongkim commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,52 @@
+
+    private static String[] splitInput(String input) {
+        String[] numbers = input.split(",|:");
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(input);
```
패턴도 미리 상수로 만들어서 관리해보시면 좋겠습니다. :)

Ganjinajae replied:
넵. 확인하였습니다.

seokhongkim commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    private static String[] splitInput(String input) {
+        String[] numbers = input.split(",|:");
```
`customDelimiter` 가 있는 경우 `numbers` 는 사용되지 않을 것 같습니다.
`numbers` 가 필요한 시점에 `split` 하는 건 어떨까요?

Ganjinajae replied:
넵. 좋은 것 같습니다.

seokhongkim commented on src/main/java/study/StringAddCalculator.java:
```diff
@@ -0,0 +1,52 @@
+        return numbers;
+    }
+
+    private static boolean isOnlyNumber(String[] numbers) {
```
숫자가 하나만 있는 경우와 2개 이상 있는 경우의 로직이 달라져야 할 필요가 있을까요?

Ganjinajae replied:
이것도 별 생각없이 기능 요구사항 힌트보다가.. 필요없는 코드가 발생한 것 같네요.
수정해보겠습니다. 👍

seokhongkim reviewed (changes requested):
영진님, 2단계 미션 잘 구현해주셨습니다. 👍 

기능 목록을 `README.md` 파일에 작성하신 후 테스트를 작성하는 과정으로 진행해보시면 학습에 도움이 될 것 같습니다. :)
이외에 코멘트 남긴 내용이 있으니 참고 부탁드립니다. 🙏 

코멘트 보시고 궁금하신 점은 깃헙 코멘트나 슬랙 DM으로 편하게 말씀 부탁드립니다. 😀

youngjin-j committed:
refactor: 코드 리뷰 반영

1. 의미를 분명하게 할 수 있게 위한 상수화
2. 필요 없는 로직 제거 (숫자 1개만 받을 때)
3. 중괄호 관련 컨벤션 통일
4. 클래스 역할 분리

youngjin-j committed:
refactor: 상수 추가

* custom 인지 아닌지를 판단하는 구분자 추가

seokhongkim commented on src/main/java/study/StringInputParser.java:
```diff
@@ -0,0 +1,25 @@
+
+public class StringInputParser {
+    public static final String DEFAULT_SPLIT_DELIMITER_REGEX = ",|:";
+    public static final String CUSTOM_SPLIT_DELIMITER_REGEX = "//(.)\n(.*)";
```
`Pattern` 은 미리 만들어 놓고 재사용해도 되고, thread safe 하므로 미리 생성해서 상수로 관리해도 괜찮을 것 같습니다. :)

Ganjinajae replied:
dm으로 문의드리겠습니다!

seokhongkim commented on src/main/java/study/StringInputParser.java:
```diff
@@ -0,0 +1,25 @@
+    public static String[] toStringArray(String input) {
+        if (input.contains(CUSTOM_DELIMITER)) {
+            Matcher m = Pattern.compile(CUSTOM_SPLIT_DELIMITER_REGEX).matcher(input);
+            if (m.find()) {
```
인덴트가 2 단계네요. 1 단계로 줄여보시면 좋겠습니다. :)

Ganjinajae replied:
다음 미션에서 반영해보겠습니다 👍

seokhongkim commented on src/main/java/study/StringInputParser.java:
```diff
@@ -0,0 +1,25 @@
+        return input.split(DEFAULT_SPLIT_DELIMITER_REGEX);
+    }
+
+    public static int toInt(String input) {
```
이 로직을 `StringInputParser` 가 제공한다면, 파싱한 결과를 바로 타입까지 변경해서 주는 것은 어떨까요?
혹은 외부에서 타입을 변경하는 책임을 가진다면, 외부에서 구현해도 될 기능으로 보입니다. :)
외부에서 구현해도 된다고 생각한 이유는 내부의 필드를 사용하지 않고, 외부에서만 사용하고 있기 때문입니다.

Ganjinajae replied:
dm으로 문의드리겠습니다!

seokhongkim reviewed (approved):
영진님, 2단계 미션 피드백 잘 반영해주셨습니다. 💯 
2단계 PR도 이만 머지하도록 하겠습니다.

처음 미션 진행하실 때 커밋을 작게 나눠주신 것처럼, 코드 리뷰 반영도 적절하게 단위를 나눠서 커밋을 작성해주시면 이해하는데 더 도움이 될 것 같습니다.

3단계 미션에서 뵙겠습니다. 😀
